spec/avram/bulk_upsert_spec.cr

```diff
@@ -0,0 +1,19 @@
+require "../spec_helper"
+
+describe Avram::BulkUpsert do
+  describe "bulk upserts" do
+    # Ideally, compiler can catch this / this should be impossible..
+    context "when collections mismatch" do
```

Comment:
Won't be able to catch this with the compiler, so we probably want to raise some sort of RunTimeError when this happens. The fields for all records need to match, otherwise you can have mismatches between the number of fields and the number of values for a given record.

Reply:
Ooh, we might be able to just pad out missing values with nil!

Reply:
I think that would overwrite to NULL for those fields.

Reply:
In an update scenario, probably -- but we could make our CONFLICT UPDATE part use coalesce to retain old values when null is passed in a bull scenario!

Reply:
ON CONFLICT (XXX) DO UPDATE SET col = COALESCE(excluded.col, col)

Reply:
Ooooh that could work ❤️

Reply:
Oh wait, that would prevent setting things to NULL, since it’d always coalesce back.

Reply:
Could you just use a named tuple as the input for each row?

Reply:
@robacarp If we used a named tuple, I think we run into a few other issues:

1. BulkUpsert class would need to be generated with macros I believe to support each class that uses it
2. You would ALWAYS need to pass ALL columns into bulk operations, which I'm not sure is something we'd want.

Reply:
Yeah the issue is mostly about the need for all keys we insert to be available in all records, otherwise some records will lose values.